frontend/src/components/TreeViewer/TreeViewer.jsx

```diff
@@ -333,37 +404,105 @@ export const TreeViewer = ({ data }) => {
+  };
 
   /**
    * This function is triggered when the home button is clicked.
    */
-  const homeClickHandler = () => historyRecorder.goBackward("");
+  const homeClickHandler = () => {
+    if (getCurrentRootName(currentPath) === getAbsoluteRootName(currentPath)) {
+      return; //No action taken if we are already at the root/home
+    }
+    const absoluteRoot = nameToNodeMapping[getAbsoluteRootName(currentPath)];
+    if (absoluteRoot == null) {
+      log("homeClickHandler: cannot find absolute root", Logger.LEVEL_ERROR);
+      return;
+    }
+    setNewVisibleRoot(absoluteRoot);
+  };
 
   /**
    * This function is triggered when the back button is clicked.
    * It goes back one level in the tree.
    */
-  const backClickHandler = () => historyRecorder.goBackward();
+  const backClickHandler = () => {
+    const previouslyVisitedNodeName = historyRecorder.goBackward(
+      getCurrentRootName(currentPath)
+    );
+    if (previouslyVisitedNodeName === "") {
+      log("backClickHandler: no previously visted node", Logger.LEVEL_WARNING);
+      return;
+    }
+    const previouslyVisitedNode = nameToNodeMapping[previouslyVisitedNodeName];
+    if (previouslyVisitedNode === null) {
+      log(
+        "backClickHandler: error finding the previously visited node given name",
+        Logger.LEVEL_WARNING
+      );
+      return;
+    }
+    setNewVisibleRoot(previouslyVisitedNode, false);
+  };
 
   /**
    * This function is triggered when the forward button is clicked.
    * It goes forward one level in the tree, if that exists.
    */
-  const forwardClickHandler = () => historyRecorder.goForward();
+  const forwardClickHandler = () => {
+    const forwardNodeName = historyRecorder.goForward(
+      getCurrentRootName(currentPath)
+    );
+    if (forwardNodeName === "") {
+      log("forwardClickHandler: no forward node", Logger.LEVEL_WARNING);
+      return;
+    }
+    const forwardNode = nameToNodeMapping[forwardNodeName];
+    if (forwardNode === null) {
+      log(
+        "forwardClickHandler: error finding the forward node given name",
+        Logger.LEVEL_WARNING
+      );
+      return;
+    }
+    setNewVisibleRoot(forwardNode, false);
+  };
 
   /**
-   * This function is triggered when an edit is made
-   * to the current path.
+   * Sets the currently hovered link which eventually triggers a
+   * re-render of the tree.
    */
-  const pathChangeHandler = (newPath) => historyRecorder.resetPath(newPath);
+  const linkHoverHandler = (hoveredElement) => {
+    if (hoveredElement === null) {
+      setHoveredNodeLink("");
+      return;
+    }
+    setHoveredNodeLink(hoveredElement.getAttribute("href"));
+  };
 
   /**
-   * This function is triggered whenever a change is made
-   * to historyRecorder.
+   * This function is triggered when an edit is made
+   * to the current path.
    */
-  const historyChangeHandler = (path) => setCurrentPath(path);
+  const pathChangeHandler = (newPath) => {
```

Comment:
I think we should also make sure that the new root has children since in nodeClickHandler(), we don't let the user click on a node that has no children, but here, the user can enter such a path. In such a case, we should just navigate to the parent. For example, if the user typed "Root/Back/CRM", we should set the new root to Back, instead of CRM.

Reply:
Added to your validatePath function to check for this case; don't think we should set the root to something different than the user typed, because that seems to be confusing and inconsistent behaviour

Reply:
That's fair